primitives/src/v2/mod.rs

```diff
@@ -1569,6 +1598,72 @@ impl CompactStatement {
 	}
 }
 
+/// `IndexedVec` struct indexed by type specific indices.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq, MallocSizeOf))]
+pub struct IndexedVec<K, V>(Vec<V>, PhantomData<fn(K) -> K>);
+
+impl<K, V> Default for IndexedVec<K, V> {
+	fn default() -> Self {
+		Self(vec![], PhantomData)
+	}
+}
+
+impl<K, V> From<Vec<V>> for IndexedVec<K, V> {
+	fn from(validators: Vec<V>) -> Self {
+		Self(validators, PhantomData)
+	}
+}
+
+impl<K, V> FromIterator<V> for IndexedVec<K, V> {
+	fn from_iter<T: IntoIterator<Item = V>>(iter: T) -> Self {
+		Self(Vec::from_iter(iter), PhantomData)
+	}
+}
+
+impl<K, V> IndexedVec<K, V>
+where
+	V: Clone,
+{
+	/// Returns a reference to an element indexed using `K`.
+	pub fn get(&self, index: K) -> Option<&V>
+	where
+		K: TypeIndex,
+	{
+		self.0.get(index.into_usize())
+	}
+
+	/// Returns number of elements in vector.
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	/// Returns contained vector.
+	pub fn to_vec(&self) -> Vec<V> {
+		self.0.clone()
+	}
+
+	/// Returns an iterator over the underlying vector.
+	pub fn iter(&self) -> Iter<'_, V> {
+		self.0.iter()
+	}
+
+	/// Returns a mutable iterator over the underlying vector.
+	pub fn iter_mut(&mut self) -> IterMut<'_, V> {
+		self.0.iter_mut()
+	}
+
+	/// Creates a consuming iterator.
+	pub fn into_iter(self) -> IntoIter<V> {
+		self.0.into_iter()
+	}
+
+	/// Returns true if the underlying container is empty.
+	pub fn is_empty(&self) -> bool {
+		self.0.is_empty()
+	}
```

Comment:
You can just implement `Deref` to get all these functions for "free"

Only `get` function is required and the compiler will then figure out the correct version to call.

Reply:
We want to ban indexing operations those and require using `.get(index)`.